force-app/main/default/classes/PSC_ManageSoftCredits_CTRL.cls

```diff
@@ -368,36 +368,102 @@ public with sharing class PSC_ManageSoftCredits_CTRL {
         
         Savepoint sp = Database.setSavepoint(); 
         try {
-            delete toDelete;
-
-            upsert toUpsertContactRoles;
-
+            Integer chunkSize = 200;
+        
+            // Step 1: Delete records that need to be removed in chunks to handle limits
+            if (!toDelete.isEmpty()) {
+                List<Partial_Soft_Credit__c> partialSoftCreditsToDelete = new List<Partial_Soft_Credit__c>();
+                List<OpportunityContactRole> contactRolesToDelete = new List<OpportunityContactRole>();
+        
+                for (SObject record : toDelete) {
+                    if (record instanceof Partial_Soft_Credit__c) {
+                        partialSoftCreditsToDelete.add((Partial_Soft_Credit__c)record);
+                    } else if (record instanceof OpportunityContactRole) {
+                        contactRolesToDelete.add((OpportunityContactRole)record);
+                    }
+                }
+        
+                // Perform deletion of Partial_Soft_Credit__c in chunks
+                for (Integer i = 0; i < partialSoftCreditsToDelete.size(); i += chunkSize) {
+                    List<Partial_Soft_Credit__c> chunk = new List<Partial_Soft_Credit__c>();
+                    for (Integer j = i; j < Math.min(i + chunkSize, partialSoftCreditsToDelete.size()); j++) {
+                        chunk.add(partialSoftCreditsToDelete[j]);
+                    }
+                    Database.DeleteResult[] deleteResults = Database.delete(chunk, false);
+                    for (Database.DeleteResult result : deleteResults) {
+                        if (!result.isSuccess()) {
+                            for (Database.Error error : result.getErrors()) {
+                                ApexPages.addMessage(new ApexPages.Message(ApexPages.Severity.ERROR, 'Delete Error (Partial Soft Credits): ' + error.getMessage()));
```

Comment:
The error handling in this implementation is not functioning correctly and needs to be rewritten. For example, if one out of ten Partial Soft Credit records encounters an error during deletion, the corresponding OpportunityContactRole record should not be deleted. You must ensure that only the related records with no DML error can be processed.

I recommend rolling back the database and displaying all errors, proceeding to the next step only if no errors are encountered.